### PR TITLE
Add RouteConfiguration and VirtualHost level configs for mirror policies

### LINKED ISSUE #19306: support request mirror policies at the virtual host and/or route configuration level
Currently you can only set request mirror policies at the route level:

https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/route/v3/route_components.proto#L1101

We have a use case where we'd like to set this for all routes in a given vhost or set of vhosts, thus it would be nice to support this at a higher scope than just route level.

### PR DESCRIPTION
**Pull request by JuniorHsu**
Fixes #19306
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message: Add RouteConfiguration and VirtualHost level configs for mirror policies
Additional Description:
Risk Level: low
Testing: unit test
Docs Changes: adjust the namespace but not amending the RouteConfigure
[Optional Fixes #Issue] #19306

### CODE REVIEW

**Comment by JuniorHsu** [comment]
The test `test/extensions/filters/http/csrf/csrf_filter_integration_test` is never passed in last 20 times and seems unrelated to this diff. Could we ignore the test failure and move on the review? @htuch @rgs1 

https://dev.azure.com/cncf/envoy/_test/analytics?definitionId=5&contextType=build

**Comment by adisuissa** [comment]
> The test `test/extensions/filters/http/csrf/csrf_filter_integration_test` is never passed in last 20 times and seems unrelated to this diff. Could we ignore the test failure and move on the review? @htuch @rgs1
> 
> https://dev.azure.com/cncf/envoy/_test/analytics?definitionId=5&contextType=build

Known issue. A fix is on the way #20262.

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:645]: This is technically change the type, since it's now top-level rather than nested in `RouteAction`. We've been weary of this in https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md, since it causes chaos in some places in general, e.g. when we have type URLs for extensions change. In this case, I think what is being done is safe and should not break any wire proto / JSON / YAML. It would potentially break some consuming code when type signature change.

My instinct is that it's relatively benign and the advantages (avoiding redundant definitions) outweigh any concerns. But would like to verify with other @envoyproxy/api-shepherds (in particular @markdroth and @mattklein123).

Comment (rgs1) [reply api/envoy/config/route/v3/route_components.proto:645]: @markdroth @mattklein123 can we get some feedback on the API please? I'll look at the rest of the impl, but would be nice to ensure we are on the same page API wise first. Thanks!

Comment (rgs1) [reply api/envoy/config/route/v3/route_components.proto:645]: Could we not use the one nested in RouteAction (with forward declaration or something similar) to avoid the controversy?

File: source/common/router/config_impl.cc
Comment (rgs1) [inline source/common/router/config_impl.cc:1450]: make_unique per above?

File: source/common/router/config_impl.cc
Comment (rgs1) [inline source/common/router/config_impl.cc:1746]: nit: i'd drop the comment.

File: api/envoy/config/route/v3/route_components.proto
Comment (rgs1) [inline api/envoy/config/route/v3/route_components.proto:1114]: Can you comment on what happens if it is specified both here and in the upper level? Do they get merged or does one of them win over the other?

**Review by rgs1 — CHANGES_REQUESTED** [review]
Thanks! Left some comments to get started.

File: api/envoy/config/route/v3/route_components.proto
Comment (JuniorHsu) [reply api/envoy/config/route/v3/route_components.proto]: That's a good point. The most internal policies win over the other. I add the comments. Thanks.

File: source/common/router/config_impl.cc
Comment (JuniorHsu) [reply source/common/router/config_impl.cc]: done

File: source/common/router/config_impl.cc
Comment (JuniorHsu) [reply source/common/router/config_impl.cc]: done

File: source/common/router/config_impl.cc
Comment (JuniorHsu) [reply source/common/router/config_impl.cc]: done

File: source/common/router/config_impl.cc
Comment (JuniorHsu) [reply source/common/router/config_impl.cc:493]: The motivation of changing pointer ownership style is to avoid deep copy while we're doing `shadow_policies_ = vhost.shadowPolicies();` The alias is already changed to `shared_ptr` as well.

Ideally we might able to alias a ptr to vector, but I suppose we wont't have a too complicated vector in the setting at this moment.

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:648]: Yes I would just leave it where it was and just reference that type. I think that's not that big of a deal and would avoid any potential compile issues for people.

Comment (JuniorHsu) [reply api/envoy/config/route/v3/route_components.proto]: I move RequestMirrorPolicy back under RouteAction to keep api unchanged. Thanks for the feedback, htuch, rgs1 and mattklein123

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:205]: You can drop all these "Inherit from.." sentences. I think they are redundant as you clearly explain the precedence relationship in the preceding sentences.

**Review by htuch — COMMENTED** [review]
API LGTM modulo docs comment.

**Review by rgs1 — CHANGES_REQUESTED** [review]
Can you add a note to `docs/root/version_history/current.rst` about the new configuration options?

File: docs/root/version_history/current.rst
Comment (JuniorHsu) [reply docs/root/version_history/current.rst:82]: Thanks for the comment.

I already tested the routeAction overwritting RouteConfiguration, and routeAction overwritting RouteConfiguration+VirtualHost.
https://github.com/envoyproxy/envoy/pull/20247/files#diff-546f801a2d2ea64fd2bf06c043801ebe5920e94744be6298f923d3e2e145141dR3662

But happy to add more like routeAction overriding VirtualHost.

File: test/common/router/config_impl_test.cc
Comment (rgs1) [inline test/common/router/config_impl_test.cc:3701]: nit: use www2 here as the name to avoid confusion

File: test/common/router/config_impl_test.cc
Comment (rgs1) [inline test/common/router/config_impl_test.cc:3680]: can you add a comment at the start of test, e.g.:

```
// Test the mirror policies are properly applied when route's don't have one and not applied when they do.
```

**Review by rgs1 — COMMENTED** [review]
Looks great! Just left a few more nits to clean up tests, other than that I am r+.

@htuch can we a final pass for API? Thanks!

File: test/common/router/config_impl_test.cc
Comment (JuniorHsu) [reply test/common/router/config_impl_test.cc]: done

File: test/common/router/config_impl_test.cc
Comment (JuniorHsu) [reply test/common/router/config_impl_test.cc]: done

**Review by rgs1 — APPROVED** [review]
Looks great, thanks!

Over to @snowp or @alyssawilk for final pass on the non api bits.

**Comment by JuniorHsu** [comment]
@htuch any chance to have one more look?
This diff needs approval from @envoyproxy/api-shepherds  
Thanks

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:202]:
```suggestion
  // Specify a set of default request mirroring policies for every route under this virtual host.
```

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by JuniorHsu** [comment]
I apply the commit suggestion in the UI but it fails DCO. Looks like I'm forced to rebase. Fortunately it's reviewed by @htuch and @rgs1 but needs another round of rubber stamps .

**Comment by JuniorHsu** [comment]
Looks this PR is broken a lot by following this instruction. The best way might be have another PR.
https://github.com/envoyproxy/envoy/pull/20247/checks?check_run_id=5580775811

**Comment by JuniorHsu** [comment]
This continues  on #20388

**Pull request closed without merging**
